File: app/src/main/java/com/khmelenko/lab/travisclient/activity/MainActivity.java
package com.khmelenko.lab.travisclient.activity;

import android.app.SearchManager;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.os.Bundle;
import android.provider.SearchRecentSuggestions;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.NavigationView;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.SearchView;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import com.khmelenko.lab.travisclient.R;
import com.khmelenko.lab.travisclient.TravisApp;
import com.khmelenko.lab.travisclient.adapter.SearchResultsAdapter;
import com.khmelenko.lab.travisclient.fragment.LicensesDialogFragment;
import com.khmelenko.lab.travisclient.fragment.ReposFragment;
import com.khmelenko.lab.travisclient.mvp.MvpActivity;
import com.khmelenko.lab.travisclient.network.response.Repo;
import com.khmelenko.lab.travisclient.network.response.User;
import com.khmelenko.lab.travisclient.presenter.RepositoriesPresenter;
import com.khmelenko.lab.travisclient.storage.SearchHistoryProvider;
import com.khmelenko.lab.travisclient.view.RepositoriesView;

import java.util.List;

import javax.inject.Inject;

import butterknife.Bind;
import butterknife.ButterKnife;

/**
 * Main application activity
 *
 * @author Dmytro Khmelenko
 */
public final class MainActivity extends MvpActivity<RepositoriesPresenter> implements RepositoriesView,
        ReposFragment.ReposFragmentListener {

    private static final int AUTH_ACTIVITY_CODE = 0;
    private static final int REPO_DETAILS_CODE = 1;

    private static final String SAVED_QUERY = "SavedQuery";

    @Bind(R.id.drawer_layout)
    DrawerLayout mDrawerLayout;

    @Inject
    RepositoriesPresenter mPresenter;

    private ReposFragment mFragment;

    private SearchView mSearchView;

    private List<String> mQueryItems;

    private String mSavedQuery;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        TravisApp.instance().activityInjector().inject(this);

        mFragment = (ReposFragment) getFragmentManager().findFragmentById(R.id.main_fragment);

        initToolbar();
        setupDrawerLayout();
    }

    @Override
    protected RepositoriesPresenter getPresenter() {
        return mPresenter;
    }

    @Override
    protected void attachPresenter() {
        getPresenter().attach(this);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString(SAVED_QUERY, mSearchView.getQuery().toString());
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        mSavedQuery = savedInstanceState.getString(SAVED_QUERY);
    }

    /**
     * Initializes toolbar
     */
    private void initToolbar() {
        final Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        final ActionBar actionBar = getSupportActionBar();

        if (actionBar != null) {
            actionBar.setHomeAsUpIndicator(R.drawable.ic_menu_white_24dp);
            actionBar.setDisplayHomeAsUpEnabled(true);
        }
    }

    /**
     * Sets up navigation drawer layout
     */
    private void setupDrawerLayout() {
        mDrawerLayout = (DrawerLayout) findViewById(R.id.drawer_layout);

        final NavigationView view = (NavigationView) findViewById(R.id.navigation_view);
        view.setNavigationItemSelectedListener(new NavigationView.OnNavigationItemSelectedListener() {
            @Override
            public boolean onNavigationItemSelected(@NonNull MenuItem menuItem) {

                switch (menuItem.getItemId()) {
                    case R.id.drawer_login:
                        Intent loginIntent = new Intent(MainActivity.this, AuthActivity.class);
                        startActivityForResult(loginIntent, AUTH_ACTIVITY_CODE);
                        break;
                    case R.id.drawer_logout:
                        getPresenter().userLogout();

                        finish();
                        startActivity(getIntent());
                        break;
                    case R.id.drawer_licenses:
                        LicensesDialogFragment dialog = LicensesDialogFragment.newInstance();
                        dialog.show(getSupportFragmentManager(), "LicensesDialog");
                        break;
                    case R.id.drawer_about:
                        Intent aboutIntent = new Intent(MainActivity.this, AboutActivity.class);
                        startActivity(aboutIntent);
                        break;
                }
                menuItem.setChecked(false);
                mDrawerLayout.closeDrawers();
                return true;
            }
        });

    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == RESULT_OK) {
            switch (requestCode) {
                case AUTH_ACTIVITY_CODE:
                    // clear previous data
                    mFragment.clearData();
                    showProgress();
                    getPresenter().reloadRepos();
                    break;
                case REPO_DETAILS_CODE:
                    boolean reloadRequired = data.getBooleanExtra(RepoDetailsActivity.RELOAD_REQUIRED_KEY, false);
                    if (reloadRequired) {
                        showProgress();
                        getPresenter().reloadRepos();
                    }
                    break;
            }
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                mDrawerLayout.openDrawer(GravityCompat.START);
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater menuInflater = getMenuInflater();
        menuInflater.inflate(R.menu.menu_main, menu);

        MenuItem searchItem = menu.findItem(R.id.action_search);
        SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);

        mSearchView = null;
        if (searchItem != null) {
            mSearchView = (SearchView) searchItem.getActionView();
        }
        if (mSearchView != null) {
            mSearchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));
            mSearchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {

                @Override
                public boolean onQueryTextSubmit(String query) {
                    // save search query to history
                    SearchRecentSuggestions suggestionsProvider = new SearchRecentSuggestions(MainActivity.this,
                            SearchHistoryProvider.AUTHORITY, SearchHistoryProvider.MODE);
                    suggestionsProvider.saveRecentQuery(query, null);
                    return false;
                }

                @Override
                public boolean onQueryTextChange(String newText) {
                    reloadSearchHistoryAdapter(newText);
                    return true;
                }

            });
            mSearchView.setOnSuggestionListener(new SearchView.OnSuggestionListener() {
                @Override
                public boolean onSuggestionClick(int position) {
                    mSearchView.setQuery(mQueryItems.get(position), true);
                    return true;
                }

                @Override
                public boolean onSuggestionSelect(int position) {
                    return true;
                }
            });

            reloadSearchHistoryAdapter("");

            // restore query if it was
            if (!TextUtils.isEmpty(mSavedQuery)) {
                mSearchView.setQuery(mSavedQuery, false);
                mSearchView.setIconified(false);
            }
        }
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public void onBackPressed() {
        if (!mSearchView.isIconified()) {
            mSearchView.setQuery("", false);
            mSearchView.setIconified(true);
        } else {
            super.onBackPressed();
        }
    }

    /**
     * Reloads search history adapter
     *
     * @param query Query
     */
    private void reloadSearchHistoryAdapter(String query) {
        Cursor cursor = SearchHistoryProvider.queryRecentSearch(this, query);
        mQueryItems = SearchHistoryProvider.transformSearchResultToList(cursor);

        mSearchView.setSuggestionsAdapter(new SearchResultsAdapter(this, cursor));
    }

    @Override
    public void onRepositorySelected(Repo repo) {
        Intent intent = new Intent(MainActivity.this, RepoDetailsActivity.class);
        intent.putExtra(RepoDetailsActivity.REPO_SLUG_KEY, repo.getSlug());
        startActivityForResult(intent, REPO_DETAILS_CODE);
    }

    @Override
    public void onRefreshData() {
        getPresenter().reloadRepos();
    }

    @Override
    public void updateUserData(User user) {
        final NavigationView view = (NavigationView) findViewById(R.id.navigation_view);
        View header = view.getHeaderView(0);
        TextView usernameView = (TextView) header.findViewById(R.id.drawer_header_username);
        TextView emailView = (TextView) header.findViewById(R.id.drawer_header_email);

        if (user != null) {
            String username = user.getLogin();
            if (!TextUtils.isEmpty(user.getName())) {
                username = String.format("%1$s (%2$s)", user.getName(), user.getLogin());
            }
            usernameView.setText(username);
            emailView.setText(user.getEmail());

            // TODO Update image, when service will provide it
        } else {
            usernameView.setText(R.string.navigation_drawer_username_placeholder);
            emailView.setText(R.string.navigation_drawer_email_placeholder);
        }
    }

    @Override
    public void setRepos(List<Repo> repos) {
        mFragment.setRepos(repos);
    }

    @Override
    public void updateMenuState(@Nullable String accessToken) {
        NavigationView view = (NavigationView) findViewById(R.id.navigation_view);
        Menu menu = view.getMenu();
        if (TextUtils.isEmpty(accessToken)) {
            menu.findItem(R.id.drawer_login).setVisible(true);
            menu.findItem(R.id.drawer_logout).setVisible(false);
        } else {
            menu.findItem(R.id.drawer_login).setVisible(false);
            menu.findItem(R.id.drawer_logout).setVisible(true);
        }
    }

    @Override
    public void showError(String message) {
        mFragment.handleLoadingFailed(message);
    }

    @Override
    public void showProgress() {
        mFragment.setLoadingProgress(true);
    }

    @Override
    public void hideProgress() {
        mFragment.setLoadingProgress(false);
    }
}

File: app-v3/src/main/java/com/khmelenko/lab/varis/repositories/MainActivity.kt
package com.khmelenko.lab.varis.repositories

import android.app.Activity
import android.app.SearchManager
import androidx.lifecycle.Observer
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.provider.SearchRecentSuggestions
import com.google.android.material.navigation.NavigationView
import androidx.core.view.GravityCompat
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.Toolbar
import android.text.TextUtils
import android.view.Menu
import android.view.MenuItem
import android.widget.TextView
import androidx.lifecycle.ViewModelProvider
import com.khmelenko.lab.varis.R
import com.khmelenko.lab.varis.about.AboutActivity
import com.khmelenko.lab.varis.about.LicensesDialogFragment
import com.khmelenko.lab.varis.activity.BaseActivity
import com.khmelenko.lab.varis.auth.AuthActivity
import com.khmelenko.lab.varis.network.response.Repo
import com.khmelenko.lab.varis.network.response.User
import com.khmelenko.lab.varis.repodetails.RELOAD_REQUIRED_KEY
import com.khmelenko.lab.varis.repodetails.REPO_SLUG_KEY
import com.khmelenko.lab.varis.repodetails.RepoDetailsActivity
import com.khmelenko.lab.varis.repositories.search.SearchResultsAdapter
import com.khmelenko.lab.varis.storage.SearchHistoryProvider
import dagger.android.AndroidInjection
import kotlinx.android.synthetic.main.activity_main.drawerLayout
import kotlinx.android.synthetic.main.activity_main.navigationView
import javax.inject.Inject

const val AUTH_ACTIVITY_CODE = 0
const val REPO_DETAILS_CODE = 1
const val SAVED_QUERY = "SavedQuery"
const val SEARCH_LIMIT = 3

/**
 * Main application activity
 *
 * @author Dmytro Khmelenko
 */
class MainActivity : BaseActivity(), ReposFragment.ReposFragmentListener {

    @Inject
    lateinit var viewModelFactory: RepositoriesViewModelFactory

    private lateinit var viewModel: RepositoriesViewModel

    private lateinit var fragment: ReposFragment

    private var searchView: SearchView? = null

    private var queryItems: List<String>? = null

    private var savedQuery: String? = null

    private lateinit var drawerToggle: ActionBarDrawerToggle

    override fun onCreate(savedInstanceState: Bundle?) {
        AndroidInjection.inject(this)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        viewModel = ViewModelProvider(this, viewModelFactory).get(RepositoriesViewModel::class.java)
        viewModel.state().observe(this, Observer { repositoriesState ->
            hideProgress()
            when (repositoriesState) {
                is RepositoriesState.Loading -> showProgress()
                is RepositoriesState.Error -> showError(repositoriesState.message)
                is RepositoriesState.ReposList -> setRepos(repositoriesState.repos)
                is RepositoriesState.UserData -> updateUserData(repositoriesState.user)
            }
        })
        viewModel.init()

        fragment = fragmentManager.findFragmentById(R.id.main_fragment) as ReposFragment

        initToolbar()
        setupDrawerLayout()

        updateMenuState(viewModel.isUserLoggedIn())
    }

    override fun onDestroy() {
        viewModel.release()
        super.onDestroy()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        if (searchView != null) {
            outState.putString(SAVED_QUERY, searchView?.query.toString())
        }
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        savedQuery = savedInstanceState.getString(SAVED_QUERY)
    }

    /**
     * Initializes toolbar
     */
    private fun initToolbar() {
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        drawerToggle = ActionBarDrawerToggle(this, drawerLayout,
                toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)

        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setHomeButtonEnabled(true)

        drawerToggle.syncState()
    }

    /**
     * Sets up navigation drawer layout
     */
    private fun setupDrawerLayout() {
        drawerLayout.addDrawerListener(drawerToggle)
        val view = findViewById<NavigationView>(R.id.navigationView)
        view.setNavigationItemSelectedListener { menuItem ->

            when (menuItem.itemId) {
                R.id.drawer_login -> {
                    val loginIntent = Intent(this@MainActivity, AuthActivity::class.java)
                    startActivityForResult(loginIntent, AUTH_ACTIVITY_CODE)
                }
                R.id.drawer_logout -> {
                    viewModel.userLogout()

                    finish()
                    startActivity(intent)
                }
                R.id.drawer_licenses -> {
                    val dialog = LicensesDialogFragment.newInstance()
                    dialog.show(supportFragmentManager, "LicensesDialog")
                }
                R.id.drawer_about -> {
                    val aboutIntent = Intent(this@MainActivity, AboutActivity::class.java)
                    startActivity(aboutIntent)
                }
            }
            menuItem.isChecked = false
            drawerLayout.closeDrawers()
            true
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
                if (resultCode == Activity.RESULT_OK) {
            when (requestCode) {
                AUTH_ACTIVITY_CODE -> {
                    // clear previous data
                    fragment.clearData()
                    showProgress()
                    viewModel.reloadRepos()
                }
                REPO_DETAILS_CODE -> {
                    val reloadRequired = data?.getBooleanExtra(RELOAD_REQUIRED_KEY, false) ?: false
                    if (reloadRequired) {
                        showProgress()
                        viewModel.reloadRepos()
                    }
                }
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                drawerLayout.openDrawer(GravityCompat.START)
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)

        val searchItem = menu.findItem(R.id.action_search)
        val searchManager = getSystemService(Context.SEARCH_SERVICE) as SearchManager

        searchView = null
        if (searchItem != null) {
            searchView = searchItem.actionView as SearchView
        }
        if (searchView != null) {
            searchView?.setSearchableInfo(searchManager.getSearchableInfo(componentName))
            searchView?.setOnQueryTextListener(object : SearchView.OnQueryTextListener {

                override fun onQueryTextSubmit(query: String): Boolean {
                    var submitProhibited = true
                    if (query.length > SEARCH_LIMIT) {
                        // save search query to history
                        val suggestionsProvider = SearchRecentSuggestions(applicationContext,
                                SearchHistoryProvider.AUTHORITY, SearchHistoryProvider.MODE)
                        suggestionsProvider.saveRecentQuery(query, null)
                        submitProhibited = false
                    }
                    return submitProhibited
                }

                override fun onQueryTextChange(newText: String): Boolean {
                    reloadSearchHistoryAdapter(newText)
                    return true
                }

            })
            searchView?.setOnSuggestionListener(object : SearchView.OnSuggestionListener {
                override fun onSuggestionClick(position: Int): Boolean {
                    searchView?.setQuery(queryItems!![position], true)
                    return true
                }

                override fun onSuggestionSelect(position: Int): Boolean {
                    return true
                }
            })

            reloadSearchHistoryAdapter("")

            // restore query if it was
            if (!TextUtils.isEmpty(savedQuery)) {
                searchView?.setQuery(savedQuery, false)
                searchView?.isIconified = false
            }
        }
        return super.onCreateOptionsMenu(menu)
    }

    override fun onBackPressed() {
        if (searchView != null && !searchView!!.isIconified) {
            searchView?.setQuery("", false)
            searchView?.isIconified = true
        } else {
            super.onBackPressed()
        }
    }

    /**
     * Reloads search history adapter
     *
     * @param query Query
     */
    private fun reloadSearchHistoryAdapter(query: String) {
        val cursor = SearchHistoryProvider.queryRecentSearch(this, query)
        queryItems = SearchHistoryProvider.transformSearchResultToList(cursor)

        searchView?.suggestionsAdapter = SearchResultsAdapter(this, cursor)
    }

    override fun onRepositorySelected(repo: Repo) {
        val intent = Intent(this@MainActivity, RepoDetailsActivity::class.java)
        intent.putExtra(REPO_SLUG_KEY, repo.slug)
        startActivityForResult(intent, REPO_DETAILS_CODE)
    }

    override fun onRefreshData() {
        viewModel.reloadRepos()
    }

    private fun updateUserData(user: User?) {
        val view = findViewById<NavigationView>(R.id.navigationView)
        val header = view.getHeaderView(0)
        val usernameView = header.findViewById<TextView>(R.id.drawer_header_username)
        val emailView = header.findViewById<TextView>(R.id.drawer_header_email)

        if (user != null) {
            var username = user.login
            if (!TextUtils.isEmpty(user.name)) {
                username = String.format("%1\$s (%2\$s)", user.name, user.login)
            }
            usernameView.text = username
            emailView.text = user.email

            // TODO Update image, when service will provide it
        } else {
            usernameView.setText(R.string.navigation_drawer_username_placeholder)
            emailView.setText(R.string.navigation_drawer_email_placeholder)
        }
    }

    private fun setRepos(repos: List<Repo>) {
        fragment.setRepos(repos)
    }

    private fun updateMenuState(userLoggedIn: Boolean) {
        val menu = navigationView.menu
        if (userLoggedIn) {
            menu.findItem(R.id.drawer_login).isVisible = false
            menu.findItem(R.id.drawer_logout).isVisible = true
        } else {
            menu.findItem(R.id.drawer_login).isVisible = true
            menu.findItem(R.id.drawer_logout).isVisible = false
        }
    }

    private fun showError(message: String?) {
        fragment.handleLoadingFailed(message)
    }

    private fun showProgress() {
        fragment.setLoadingProgress(true)
    }

    fun hideProgress() {
        fragment.setLoadingProgress(false)
    }
}
